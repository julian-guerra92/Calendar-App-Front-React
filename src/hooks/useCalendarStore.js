import { useDispatch, useSelector } from 'react-redux';
import Swal from 'sweetalert2';
import calendarApi from '../api/calendarApi';
import { covertEventsToDateEvents } from '../helpers';
import { onAddnewEvent, onDeleteEvent, onLoadEvents, onSetActiveEvent, onUpdateEvent } from '../store';

export const useCalendarStore = () => {

   const dispatch = useDispatch();

   const { events, activeEvent } = useSelector(state => state.calendar);

   const { user } = useSelector(state => state.auth);

   const setActiveEvent = (calendarEvent) => {
      dispatch(onSetActiveEvent(calendarEvent));
   }

   const startSavinfEvent = async (calendarEvent) => {
      try {
         if (calendarEvent.id) {
            //*Actualizando evento
            await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent)
            dispatch(onUpdateEvent({ ...calendarEvent, user }));
            return;
         }
         //*Creando Evento
         const { data } = await calendarApi.post('/events', calendarEvent);
         console.log(data);
         dispatch(onAddnewEvent({ ...calendarEvent, id: data.saveEvent.id, user }));
      } catch (error) {
         console.log(error);
         Swal.fire('Error to Save', error.response.data.msg, 'error');
      }
   }

   const startDeletingEvent = async () => {
      try {
         await calendarApi.delete(`/events/${activeEvent.id}`)
         dispatch(onDeleteEvent());
      } catch (error) {
         console.log(error);
         Swal.fire('Error to Delete', error.response.data.msg, 'error');
      }
   }

   const startLoadingEvents = async () => {
      try {
         const { data } = await calendarApi.get('/events');
         const events = covertEventsToDateEvents(data.events);
         dispatch(onLoadEvents(events));
      } catch (error) {
         console.log('Error cargando eventos');
         console.log(error);
      }
   }

   return {
      //*Properties
      activeEvent,
      events,
      hasEventSelected: !!activeEvent,
      //*Methods
      setActiveEvent,
      startDeletingEvent,
      startLoadingEvents,
      startSavinfEvent,
   }
}
