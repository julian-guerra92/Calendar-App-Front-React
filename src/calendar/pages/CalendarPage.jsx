import { useEffect, useState } from 'react';
import { Calendar } from 'react-big-calendar';
import { CalendarEvent, NavBar, CalendarModal, FabAddNew, FabbDelete } from '../';
import { localizer, getMessagesES } from '../../helpers';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import { useAuthStore, useCalendarStore, useUiStore } from '../../hooks';

export const CalendarPage = () => {

   const { user } = useAuthStore();

   const [lenguage, setLenguage] = useState(false);

   const [lastView, setLastView] = useState(localStorage.getItem('lastView') || 'week');

   const { openDateModal } = useUiStore();

   const { events, setActiveEvent, startLoadingEvents } = useCalendarStore();

   const onChangeLenguage = () => {
      setLenguage(current => !current);
   }

   const eventStyleGetter = (event, start, end, isSelected) => {
      const isMyEvent = (user.uid === event.user._id) || (user.uid === event.user.uid);
      const style = {
         backgroundColor: isMyEvent? '#347CF7' : '#465660',
         borderRadius: '0px',
         opacity: 0.8,
         color: 'white'
      }
      return {
         style
      }
   }

   const onDoubleClick = (event) => {
      openDateModal();
   }

   const onSelect = (event) => {
      setActiveEvent(event);
   }

   const onViewChanged = (event) => {
      localStorage.setItem('lastView', event);
      setLastView(event);
   }

   useEffect(() => {
      startLoadingEvents();
   }, []);

   return (
      <>
         <NavBar onChangeLenguage={onChangeLenguage} />
         <Calendar
            culture={lenguage && 'es'}
            messages={lenguage && getMessagesES()}
            localizer={localizer}
            events={events}
            defaultView={lastView}
            startAccessor="start"
            endAccessor="end"
            style={{ height: 'calc(100vh - 80px)' }}
            eventPropGetter={eventStyleGetter}
            components={{
               event: CalendarEvent
            }}
            onDoubleClickEvent={onDoubleClick}
            onSelectEvent={onSelect}
            onView={onViewChanged}
         />
         <CalendarModal />
         <FabAddNew />
         <FabbDelete />
      </>
   )
}
