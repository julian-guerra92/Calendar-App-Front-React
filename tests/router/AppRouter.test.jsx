import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { CalendarPage } from '../../src/calendar/pages/CalendarPage';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/calendar/pages/CalendarPage', () => ({
   CalendarPage: () => <h1>Calendar Page</h1>
}))


describe('Pruebas en <AppRouter/>', () => {

   const mockCheckAuthToken = jest.fn();

   beforeEach(() => jest.clearAllMocks());

   test('Debe de mostrar la pantalla de carga y llamar el checkAuthToken', () => {
      useAuthStore.mockReturnValue({
         status: 'checking',
         checkAuthToken: mockCheckAuthToken,
      })
      render(<AppRouter />);
      expect(screen.getByText('Cargando...')).toBeTruthy();
      expect(mockCheckAuthToken).toHaveBeenCalled();
   })

   test('Debe de mostrar el login en caso de no estar autenticado', () => {
      useAuthStore.mockReturnValue({
         status: 'not-authenticated',
         checkAuthToken: mockCheckAuthToken,
      })
      const { container } = render(
         <MemoryRouter>
            <AppRouter />
         </MemoryRouter>
      );
      expect(screen.getAllByText('Sing In')).toBeTruthy();
      expect(container).toMatchSnapshot();
   })

   test('Debe de mostrar el calendario si estamos autenticados', () => {
      useAuthStore.mockReturnValue({
         status: 'authenticated',
         checkAuthToken: mockCheckAuthToken,
      })
      render(
         <MemoryRouter>
            <AppRouter />
         </MemoryRouter>
      );
      expect(screen.getByText('Calendar Page')).toBeTruthy();
   })

})