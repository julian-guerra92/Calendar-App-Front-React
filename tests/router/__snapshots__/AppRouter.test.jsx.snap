// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter/> Debe de mostrar el login en caso de no estar autenticado 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Sing In
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Sing In"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          User Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Full Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create Count"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
